components/validator/guardian.go

```diff
@@ -103,121 +91,101 @@ func (g *Guardian) ValidateL2Output(ctx context.Context, outputRoot eth.Bytes32,
+	needConfirm, err := g.checkConfirmCondition(ctx, event.TransactionId)
 	if err != nil {
-		return nil, err
+		// error occurred. go to retry.
+		return err
 	}
-	if executionTx.Executed {
-		log.Warn("transaction %d already executed.", transactionId)
-		return nil, nil
+	if !needConfirm {
+		return nil
 	}
 
-	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
-	executionGas, err := g.cfg.L1Client.EstimateGas(cCtx, ethereum.CallMsg{
-		From:  g.cfg.SecurityCouncilAddr,
-		To:    &executionTx.Destination,
-		Value: executionTx.Value,
-		Data:  executionTx.Data,
-	})
-	cCancel()
+	isValid, err := g.ValidateL2Output(ctx, event.OutputRoot, event.L2BlockNumber.Uint64())
 	if err != nil {
-		return nil, fmt.Errorf("failed to estimate confirmation tx gas: %w", err)
+		return fmt.Errorf("failed to verify the output to be replaced. (transactionId: %d): %w", event.TransactionId.Int64(), err)
 	}
+	if isValid {
+		g.log.Info("the output to be replaced is valid")
 
-	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
-	txOpts := utils.NewSimpleTxOpts(cCtx, g.cfg.TxManager.From(), g.cfg.TxManager.Signer)
-	confirmTx, err := g.securityCouncilContract.ConfirmTransaction(txOpts, transactionId)
+		cCtx, cCancel := context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+		tx, err := g.ConfirmTransaction(cCtx, event.TransactionId)
+		cCancel()
+		if err != nil {
+			return fmt.Errorf("failed to create confirm tx. (transactionId: %d): %w", event.TransactionId.Int64(), err)
+		}
+		if tx == nil {
+			return nil
+		}
+
+		if txResponse := g.cfg.TxManager.SendTransaction(ctx, tx); txResponse.Err != nil {
+			return fmt.Errorf("failed to send confirm tx. (transactionId: %d): %w", event.TransactionId.Int64(), txResponse.Err)
+		}
+	} else {
+		g.log.Info("ignore this challenge because the output to be replaced is invalid")
+	}
+	return nil
+}
+
+func (g *Guardian) checkConfirmCondition(ctx context.Context, transactionId *big.Int) (bool, error) {
+	cCtx, cCancel := context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+	callOpts := utils.NewCallOptsWithSender(cCtx, g.cfg.TxManager.From())
+	isConfirmed, err := g.securityCouncilContract.IsConfirmed(callOpts, transactionId)
 	cCancel()
 	if err != nil {
-		return nil, err
+		return true, fmt.Errorf("failed to get confirmation. (transactionId: %d): %w", transactionId.Int64(), err)
+	}
+	if isConfirmed {
+		return false, nil
 	}
 
 	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
-	confirmationGas, err := g.cfg.L1Client.EstimateGas(cCtx, ethereum.CallMsg{
-		From:  txOpts.From,
-		To:    confirmTx.To(),
-		Data:  confirmTx.Data(),
-		Value: confirmTx.Value(),
-	})
+	executionTx, err := g.securityCouncilContract.Transactions(&bind.CallOpts{Context: cCtx}, transactionId)
 	cCancel()
 	if err != nil {
-		return nil, fmt.Errorf("failed to estimate confirmation tx gas: %w", err)
+		return true, fmt.Errorf("failed to get transaction with id: %d): %w", transactionId.Int64(), err)
```

Comment:
```suggestion
		return false, fmt.Errorf("failed to get transaction with id: %d: %w", transactionId.Int64(), err)
```

Reply:
The preceding bool value is meaningless when the error return value is prioritized, but i assumed the default value was `true` because in this case need to retry the confirm action.